### PR TITLE
NXP-22949: Add the ability to use a server-side KMS managed key for S3 encryption

### PR DESCRIPTION
**Pull request by csetera**
This pull request adds the ability to the S3 binary manager to use a KMS-managed key for server-side encryption.  It adds a new crypt.kms.key property to configure the KMS key's identifier to be used.  When coupled with proper IAM roles, this makes for a very secure and easy to manage encryption setup.

### CODE REVIEW

**Comment by efge** [comment]
Thanks for your pull request. You'll need to sign the CLA above for us to move forward.

When this is done, could you update the commit message to prefix it with "NXP-22949: " as we have this issue [NXP-22949](https://jira.nuxeo.com/browse/NXP-22949) that describes it. Also I'm make a few cosmetic comments on the code itself.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 1]
```diff
@@ -137,7 +138,9 @@ protected String getDefaultDigestAlgorithm() {
     public static final String KEYSTORE_PASS_PROPERTY = "crypt.keystore.password";
 
     public static final String SERVERSIDE_ENCRYPTION_PROPERTY = "crypt.serverside";
-
+
```
Comment (efge) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: You've added trailing spaces here (and in other places), please check that this doesn't happen.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 2]
```diff
@@ -164,8 +167,9 @@ protected String getDefaultDigestAlgorithm() {
 
     protected CryptoConfiguration cryptoConfiguration;
 
-    protected boolean userServerSideEncryption;
-
+    protected boolean useServerSideEncryption;
+    protected String serverSideKMSKeyID;
```
Comment (efge) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Please keep a blank line between two fields.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 3]
```diff
@@ -445,9 +450,16 @@ public void storeFile(String digest, File file) throws IOException {
                 PutObjectRequest request;
                 if (!isEncrypted) {
                     request = new PutObjectRequest(bucketName, key, file);
-                    if (userServerSideEncryption) {
+                    if (useServerSideEncryption) {
                         ObjectMetadata objectMetadata = new ObjectMetadata();
-                        objectMetadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                        if (serverSideKMSKeyID != null) {
+                        		SSEAwsKeyManagementParams keyManagementParams =
```
Comment (efge) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Please don't use tabs but 4 spaces for indent.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 4]
```diff
@@ -445,9 +450,16 @@ public void storeFile(String digest, File file) throws IOException {
                 PutObjectRequest request;
                 if (!isEncrypted) {
                     request = new PutObjectRequest(bucketName, key, file);
-                    if (userServerSideEncryption) {
+                    if (useServerSideEncryption) {
                         ObjectMetadata objectMetadata = new ObjectMetadata();
-                        objectMetadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                        if (serverSideKMSKeyID != null) {
```
Comment (efge) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Replace `serverSideKMSKeyID != null` with `isNotBlank(serverSideKMSKeyID)` as this comes from user input and we prefer to be lenient with spaces in that case.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 5]
```diff
@@ -445,9 +450,16 @@ public void storeFile(String digest, File file) throws IOException {
...
+                        		SSEAwsKeyManagementParams keyManagementParams = 
+                    				new SSEAwsKeyManagementParams(serverSideKMSKeyID);
+                        		request = request.withSSEAwsKeyManagementParams(keyManagementParams);
+                        } else {
+                        		objectMetadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                        }
+
```
Comment (efge) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Remove unneeded blank line.

**Comment by csetera** [comment]
I'm checking on the CLA with my company.  Once I've cleared that, I will do whatever you need done.  In the case of the review comments, have those already been applied or do I need to make those changes and resubmit?

**Comment by efge** [comment]
For the review comments it would be nice if you could amend your commit and just force-push on your branch. The PR will be updated automatically.

**Comment by csetera** [comment]
I've signed the CLA and *attempted* to update the code per your requests.  Please let me know if I did or did not get that right.

**Comment by efge** [comment]
Thank you, this looks fine. We'll do a bit of QA on this and merge if all is well.

**Comment by efge** [comment]
Hm I'm not seeing the CLA as being signed. This seems due to the fact that your commit message doesn't use your GitHub email. Could you check that please @csetera ?

**Comment by csetera** [comment]
I added my work email to my account.  I believe that was what was linked to the commit.  Did that clear things up?

**Comment by efge** [comment]
Perfect thanks.

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 6]
```diff
@@ -138,6 +139,8 @@ protected String getDefaultDigestAlgorithm() {
 
     public static final String SERVERSIDE_ENCRYPTION_PROPERTY = "crypt.serverside";
 
+    public static final String SERVERSIDE_ENCRYPTION_KMS_KEY_PROPERTY = "crypt.kms.key";
```
Comment (driverpt) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: You should rename this to `crypt.kms.key.id` so it's consistent with the Field

File: nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java [hunk 7]
```diff
@@ -224,7 +229,8 @@ protected void setupCloudClient() throws IOException {
         String endpoint = getProperty(ENDPOINT_PROPERTY);
         String sseprop = getProperty(SERVERSIDE_ENCRYPTION_PROPERTY);
         if (isNotBlank(sseprop)) {
-            userServerSideEncryption = Boolean.parseBoolean(sseprop);
```
Comment (driverpt) [inline nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Can you put this into a separate commit, like `Fix Typo` ?

Comment (efge) [reply nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/S3BinaryManager.java]: Ideally yes but it's a small enough fix that this doesn't matter here.

**Comment by csetera** [comment]
So, where are we?  What do I need to do to get this to a point where you are able to accept?

**Comment by efge** [comment]
As I mentioned let us do some additional QA/testing on our side and we'll merge it.

**Comment by csetera** [comment]
Works for us.  We have our own build, so it isn't time crucial.  I just wanted to make sure you have what you need.

**Comment by efge** [comment]
Now merged. [NXP-22949](https://jira.nuxeo.com/browse/NXP-22949) will have details.
